### PR TITLE
fix: fixes datadog tags overwritten

### LINKED ISSUE #41: Pattern Basic overwrites datadog tag environment variables
# This is a Bug Report

## Description

* What went wrong?
After switching to DAZN internal async Datadog library custom metrics were lost.  The cause was shown to be a change in tags used due to `process.env.DATADOG_TAGS` being overwritten.

* What did you expect should have happened?
No change to the tags that were configured.

* What was the config (if any) you used?
Tag `env:{ENV}` was changed to `environment:{ENV}`

* What stacktrace or error message did you see?
N/A

Similar or dependent issues:
N/A

## Additional Data

* ***Which powertool package are you using?***: lambda-powertools-pattern-basic
* ***Which version are you using?***: 1.1.4
* ***Middy engine version you're using***: 0.15.1
* ***Operating System***: N/A
* ***Stack Trace***: N/A
* ***Provider Error messages***: N/A

### PR DESCRIPTION
**Pull request by ChrisWestcottUK**
## What did you implement:

Closes #41 

Doesn't overwrite `process.env.DATADOG_TAGS`

## How did you implement it:

Only assigns `process.env.DATADOG_TAGS` if they have not been assigned

## How can we verify it:

This is a very trivial bug fix and so hopefully this is out of scope.

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [ ] Provide verification config / projects / resources
- [ ] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** Yes
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by theburningmonk** [comment]
@ChrisWestcottUK instead of an either-or, can we make it so that it complements existing DATADOG_TAGS instead? I guess that means having to pass the CSV string if one is already there, and then add properties to it, and then CSV the whole thing.

**Comment by ChrisWestcottUK** [comment]
@theburningmonk - yes I could do that, however, I do wonder whether this should add any datadog tags at all.

Although we likely have dashboards that rely on its presence in this package, isn't datadog really out of scope of the lambda power tools?  Especially if viewed as an open source library now.

**Comment by theburningmonk** [comment]
@ChrisWestcottUK that's a good point. Esp from the open source perspective, this assumption on DataDog is no longer valid, but a lot of our internal users would suddenly find their dashboards, etc. not working because of missing tags. That's a bigger, breaking change than I'd like to make.

The tagging format might be based on DataDog but it's still useful and easily parsed and transformed for other vendors too. So I still think they're worth having.

File: packages/lambda-powertools-pattern-basic/index.js [hunk 1]
```diff
@@ -12,7 +12,24 @@ if (!process.env.DATADOG_PREFIX) {
...
+  const existingNormalised = existing.split(',').map(pair => pair.split(':'))
+
+  // Assigning to a map to stop any duplicates keys (existing taking precedence)
+  const allTags = new Map([...toAdd, ...existingNormalised])
+
+  // convert back to original csv format
+  return Array.from(allTags).map(i => i.join(':')).join(',')
+}
+
+process.env.DATADOG_TAGS = supplementCsv(process.env.DATADOG_TAGS, [
```
Comment (gsingh1) [inline packages/lambda-powertools-pattern-basic/index.js]: What if this env is not set? `split` on this would error above I believe.

Comment (gsingh1) [reply packages/lambda-powertools-pattern-basic/index.js]:
```suggestion
process.env.DATADOG_TAGS = supplementCsv(process.env.DATADOG_TAGS || '', [
```

File: packages/lambda-powertools-pattern-basic/index.js [hunk 2]
```diff
@@ -12,7 +12,24 @@ if (!process.env.DATADOG_PREFIX) {
   process.env.DATADOG_PREFIX = FUNCTION_NAME + '.'
 }
 
-process.env.DATADOG_TAGS = `awsRegion:${AWS_REGION},functionName:${FUNCTION_NAME},functionVersion:${FUNCTION_VERSION},environment:${ENV}`
+const supplementCsv = (existing, toAdd) => {
+  // existing will be in the format '<key1>:<value1>,<key2>:<value2>'
+  // Map requires as [[key1, value1], [key2, value2]]
+  const existingNormalised = existing.split(',').map(pair => pair.split(':'))
```
Comment (gsingh1) [inline packages/lambda-powertools-pattern-basic/index.js]:
```suggestion
  const existingNormalised = existing && existing.split(',').map(pair => pair.split(':')) || []
```

Comment (ChrisWestcottUK) [reply packages/lambda-powertools-pattern-basic/index.js]: @gsingh1  - I've reworked this change considerably, which should not only address this but also make using supplementing the csv data easier to use if we pull it into the other patterns

**Comment by theburningmonk** [comment]
@ChrisWestcottUK give me a shout when you're done with your changes and ready for review again

**Comment by ChrisWestcottUK** [comment]
@theburningmonk feel free to look at this change

File: packages/lambda-powertools-pattern-basic/supplement-csv.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const csvToKeyValuesArray = csv =>
+  csv.split(',').filter(x => x).map(pair => pair.split(':'))
```
Comment (theburningmonk) [inline packages/lambda-powertools-pattern-basic/supplement-csv.js]: the second `split` need to be a bit smarter and deal with cases like `env:stage:test` (datadog't [official doc](https://docs.datadoghq.com/tagging/) mentions this example too)

image

File: packages/lambda-powertools-pattern-basic/index.js [hunk 4]
```diff
@@ -12,7 +13,10 @@ if (!process.env.DATADOG_PREFIX) {
   process.env.DATADOG_PREFIX = FUNCTION_NAME + '.'
 }
 
-process.env.DATADOG_TAGS = `awsRegion:${AWS_REGION},functionName:${FUNCTION_NAME},functionVersion:${FUNCTION_VERSION},environment:${ENV}`
+process.env.DATADOG_TAGS = supplementCsv({
+  existing: process.env.DATADOG_TAGS,
+  additional: `awsRegion:${AWS_REGION},functionName:${FUNCTION_NAME},functionVersion:${FUNCTION_VERSION},environment:${ENV}`
```
Comment (theburningmonk) [inline packages/lambda-powertools-pattern-basic/index.js]: since we're in control of the `additional` tags, would it be easier if we pass in a map instead of a CSV string?

Comment (theburningmonk) [reply packages/lambda-powertools-pattern-basic/index.js]: e.g. 
```
process.env.DATADOG_TAGS = supplementCsv({
  existing: process.env.DATADOG_TAGS,
  additional: {
    awsRegion: AWS_REGION,
    functionName: FUNCTION_NAME,
    functionVersion: FUNCTION_VERSION,
    environment: ENV
  }
})
```

File: packages/lambda-powertools-pattern-basic/supplement-csv.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+const keyValuesArrayToCsv = pairs =>
+  pairs.map(x => x.join(':')).join(',')
+
+const combineUniqueKeyValueArray = (original, additional) =>
```
Comment (theburningmonk) [inline packages/lambda-powertools-pattern-basic/supplement-csv.js]: `existing` instead of `original` so we name things consistently

Comment (theburningmonk) [reply packages/lambda-powertools-pattern-basic/supplement-csv.js]: Also, when we call this function on ln17, we pass in `additionalNormalised` first, which we call `original` here. Is that not the wrong way round?

File: packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      }
+
+      const actual = supplementCsv(inputs)
+
+      expect(actual).toEqual(expected)
+    })
+  })
+
+  describe('when there is an existing csv string', () => {
+    it('should supplement with items to add', () => {
```
Comment (theburningmonk) [inline packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: this test is quite hard to follow, why split them up? why not just expect the output of supplementCSV to equal `key1:value1,key2:value2,key3:value3,key4:value4,key5:value5,key6:value6`?

File: packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+      actual.forEach(pair => {
+        expect(expected).toContainEqual(pair)
+      })
+
+      expect(actual.length).toEqual(expected.length)
+    })
+
+    it('should not overwrite existing keys with items to add', () => {
+      const expected = 'key1:value1,key2:value2,key3:value3,key5:value5,key6:value6'.split(',')
```
Comment (theburningmonk) [inline packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: as above, why split the `expected` and make me compute and hold the result in my head?

File: packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+      const actual = supplementCsv(inputs).split(',')
+
+      actual.forEach(pair => {
+        expect(expected).toContainEqual(pair)
+      })
+
+      expect(actual.length).toEqual(expected.length)
+    })
+  })
```
Comment (theburningmonk) [inline packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: Need another test case to make sure we handle inputs like `env:stage:test` (see below), and when the tag is just a `value` not `key:value`

Comment (theburningmonk) [reply packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: datadog supports both `value` and `key:value` for tags

image

Comment (ChrisWestcottUK) [reply packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: That makes things interesting. Very good spot!

File: packages/lambda-powertools-pattern-basic/index.js (hunk 4 above)
Comment (ChrisWestcottUK) [reply packages/lambda-powertools-pattern-basic/index.js]: The reason for having it as a csv string was to minimise the impact of change if it is rolled out to other patterns...  but it is much easier to construct a map than build a string so I'm happy to make that change.

File: packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js (hunk 6 above)
Comment (ChrisWestcottUK) [reply packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: I'm not guaranteeing the order of the returned csv as it is fully reconstructed.

Although the order could be worked out, I don't think we should rely on the implementation details of the function under test, otherwise the test will become brittle.

**Comment by ChrisWestcottUK** [comment]
@theburningmonk - would you like another look at this now?

File: packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js (hunk 6 above)
Comment (theburningmonk) [reply packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: I see, how about using `expect(...).to.deepEqual(...)` and compare the actual and expected after we split them by `,`?

Comment (ChrisWestcottUK) [reply packages/lambda-powertools-pattern-basic/__tests__/supplement-csv.js]: I might be missing something here. As this is an array the order of the elements is important:

```
expect([1,2,3]).toEqual([2,1,3]) => fail
```

**Comment by ChrisWestcottUK** [comment]
Thank you @theburningmonk and @gsingh1 for taking the time to look at this
